# Analyzing "Chinook Music Store" With SQL
Chinook database is a media-related sample database in SQLite used for practicing purpose. In this project I will use chinook database to achieve the following objectives:

- Find the most popular and the lesat popular genres
- Investigate the most popular genres by country
- Evaluate the performance of the sales support agents
- Identify the portions of whole album purchase and individual track purchase


In [5]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

In [6]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table", "view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

#### Inspecting all the tables:

In [7]:
%%sql 
SELECT * FROM album LIMIT 3

Done.


[(1, 'For Those About To Rock We Salute You', 1),
 (2, 'Balls to the Wall', 2),
 (3, 'Restless and Wild', 2)]

In [8]:
%%sql 
SELECT * FROM artist LIMIT 3

Done.


[(1, 'AC/DC'), (2, 'Accept'), (3, 'Aerosmith')]

In [9]:
%%sql 
SELECT * FROM customer LIMIT 3

Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3)]

In [10]:
%%sql 
SELECT * FROM employee LIMIT 3

Done.


[(1, 'Adams', 'Andrew', 'General Manager', None, '1962-02-18 00:00:00', '2016-08-14 00:00:00', '11120 Jasper Ave NW', 'Edmonton', 'AB', 'Canada', 'T5K 2N1', '+1 (780) 428-9482', '+1 (780) 428-3457', 'andrew@chinookcorp.com'),
 (2, 'Edwards', 'Nancy', 'Sales Manager', 1, '1958-12-08 00:00:00', '2016-05-01 00:00:00', '825 8 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 2T3', '+1 (403) 262-3443', '+1 (403) 262-3322', 'nancy@chinookcorp.com'),
 (3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com')]

In [11]:
%%sql 
SELECT * FROM genre LIMIT 3

Done.


[(1, 'Rock'), (2, 'Jazz'), (3, 'Metal')]

In [12]:
%%sql 
SELECT * FROM invoice LIMIT 3

Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84),
 (2, 30, '2017-01-03 00:00:00', '230 Elgin Street', 'Ottawa', 'ON', 'Canada', 'K2P 1L7', 9.9),
 (3, 40, '2017-01-05 00:00:00', '8, Rue Hanovre', 'Paris', 'None', 'France', '75002', 1.98)]

In [13]:
%%sql 
SELECT * FROM invoice_line LIMIT 3 

Done.


[(1, 1, 1158, 0.99, 1), (2, 1, 1159, 0.99, 1), (3, 1, 1160, 0.99, 1)]

In [14]:
%%sql 
SELECT * FROM media_type LIMIT 3

Done.


[(1, 'MPEG audio file'),
 (2, 'Protected AAC audio file'),
 (3, 'Protected MPEG-4 video file')]

In [15]:
%%sql 
SELECT * FROM playlist LIMIT 3

Done.


[(1, 'Music'), (2, 'Movies'), (3, 'TV Shows')]

In [16]:
%%sql
SELECT * FROM playlist_track LIMIT 3

Done.


[(1, 3402), (1, 3389), (1, 3390)]

In [17]:
%%sql 
SELECT * FROM track LIMIT 3

Done.


[(1, 'For Those About To Rock (We Salute You)', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 343719, 11170334, 0.99),
 (2, 'Balls to the Wall', 2, 2, 1, None, 342562, 5510424, 0.99),
 (3, 'Fast As a Shark', 3, 2, 1, 'F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman', 230619, 3990994, 0.99)]

### Objective 1: Find the most popular and the least popular genres

Identifying the most popular and the least popular genres can help the store decides which genres of albums to put on the shelf, and which genres to stop purchasing. 

The popularity of a genre can be reflected by the number of tracks sold in this genre. Based on the table information above, the *invoice_line* table contains the information of the number of tracks sold, and the *genre* table contains the information of genre. Combine both information, the most and the least popular genres can be found.

In [26]:
%%sql

WITH tg AS
    (
    SELECT
        track_id,
        genre_id
    FROM track
    )
SELECT 
    g.name genre, 
    SUM(i.quantity) tracks_sold
FROM genre g
LEFT JOIN tg ON g.genre_id == tg.genre_id
LEFT JOIN invoice_line i ON i.track_id == tg.track_id
GROUP BY 1
ORDER BY 2 DESC

Done.


[('Rock', 2635),
 ('Metal', 619),
 ('Alternative & Punk', 492),
 ('Latin', 167),
 ('R&B/Soul', 159),
 ('Blues', 124),
 ('Jazz', 121),
 ('Alternative', 117),
 ('Easy Listening', 74),
 ('Pop', 63),
 ('Electronica/Dance', 55),
 ('Classical', 47),
 ('Reggae', 35),
 ('Hip Hop/Rap', 33),
 ('Heavy Metal', 8),
 ('Soundtrack', 5),
 ('TV Shows', 2),
 ('Drama', 1),
 ('Bossa Nova', None),
 ('Comedy', None),
 ('Opera', None),
 ('Rock And Roll', None),
 ('Sci Fi & Fantasy', None),
 ('Science Fiction', None),
 ('World', None)]

As shown in the table above, the top 3 popular genres are Rock, Metal, Alternative & Punk. Interestingly, Rock music stands out as the most popular genre and has over four times more sales than Metal music, which is the second most popular genre, and over five times more sells than Alternative & Punk music, the third most popular music. Latin and R&B music, even though they are also in the top 5 most popular genres list, have a much lower quantity of tracks sold than the top 3. Besides, there is not a huge difference between Latin, the 4th most popular genre, and Alternative, the 8th most popular genre.

There are many genres which barely, or even do not have any sale at all. These include Heavy Metal, Soundtrack, TV Shows, Drama, Bossa Nova, Opera, Rock And Roll, Sci Fi & Fantasy, Science Fiction, and World. Interestingly, Rock music is dominant in sales while Rock And Roll music has no sale. As shown in the table below, the reason is possible because BackBeat is not popular in Chinook, as only the tracks by BackBeat belong to Rock And Roll. 

Based on the result above, **Rock, Metal and Alternative & Puck music** are the best options to boost the sales volume. The other great choices are Latin, R&B/Soul, Blues, Jazz, and Alternative music. The store should avoid purchasing more tracks in Heavy Metal, Soundtrack, TV Shows, Drama, Bossa Nova, Opera, Rock And Roll, Sci Fi & Fantasy, Science Fiction, and World, as they have very low sales. 

In [32]:
%%sql
SELECT DISTINCT
    ar.name artist_name,
    g.name genre
FROM genre g
INNER JOIN track t ON g.genre_id == t.genre_id
INNER JOIN album al ON al.album_id == t.album_id
INNER JOIN artist ar ON ar.artist_id == al.artist_id
WHERE g.name == 'Rock And Roll'

Done.


[('BackBeat', 'Rock And Roll')]

### Objective 2: Investigate the most popular genres by country

It is common that people from the same country have similar tastes in music. Therefore, understand the most popular genres by country can help the sales support agents recommend the music that a new customer potentially likes once they know about the customer's nationality. 

From the tables, there are two columns that indicate the customer's nationality: the "country" column from the *customer* table and the "billing_country" column from the *invoice* table. I need to compare if these two columns have the same information. Other than the country information, I also need the genre and the quantity of purchase for each genre in each country. 

In [19]:
%%sql
SELECT
    c.customer_id,
    c.country,
    i.billing_country
FROM customer c
INNER JOIN invoice i ON i.customer_id == c.customer_id
WHERE c.country <> i.billing_country

Done.


[]

The above table is empty, indicating that the "country" column from the *customer* table and the "billing_country" column from the *invoice* table are the same. So I will just use the "country" column from the *customer* table.

In [20]:
%%sql
WITH t AS
    (
    SELECT
        track_id,
        genre_id
    FROM track
    ),
    il AS
    (
    SELECT 
        track_id,
        invoice_id,
        quantity
    FROM invoice_line
    ),
    i AS
    (
    SELECT
        invoice_id,
        customer_id
    FROM invoice
    ),
    c AS
    (
    SELECT 
        customer_id,
        country
    FROM customer
    ),
    total AS    
(SELECT
    c.country country,
    g.name genre,
    SUM(il.quantity) total_purchases
FROM genre g
INNER JOIN t ON g.genre_id == t.genre_id
INNER JOIN il ON il.track_id == t.track_id
INNER JOIN i ON i.invoice_id == il.invoice_id
INNER JOIN c ON c.customer_id == i.customer_id
GROUP BY 1,2
)

SELECT * FROM total
GROUP BY 1
HAVING total_purchases == MAX(total_purchases)

Done.


[('Argentina', 'Alternative & Punk', 17),
 ('Australia', 'Rock', 34),
 ('Austria', 'Rock', 40),
 ('Belgium', 'Rock', 26),
 ('Brazil', 'Rock', 205),
 ('Canada', 'Rock', 333),
 ('Chile', 'Rock', 61),
 ('Czech Republic', 'Rock', 143),
 ('Denmark', 'Rock', 24),
 ('Finland', 'Rock', 46),
 ('France', 'Rock', 211),
 ('Germany', 'Rock', 194),
 ('Hungary', 'Rock', 44),
 ('India', 'Rock', 102),
 ('Ireland', 'Rock', 72),
 ('Italy', 'Rock', 35),
 ('Netherlands', 'Rock', 33),
 ('Norway', 'Rock', 40),
 ('Poland', 'Rock', 40),
 ('Portugal', 'Rock', 108),
 ('Spain', 'Rock', 46),
 ('Sweden', 'Rock', 60),
 ('USA', 'Rock', 561),
 ('United Kingdom', 'Rock', 166)]

Surprisingly, the favorite genre for almost all countries is Rock. Only Argentina people like Alternative & Punk the most. Therefore, the sales support agents may recommend Rock music for most of the new customers, if they do not have any specific music to purchase in mind. 

### Objective 3: Evaluate the performance of the sales support agents

In Chinook store, every customer gets assigned to a sales support agent when they first make a purchase. My second task is to evaluate the performance of the sales support agents. 

The performance of a sales support agent is directly linked to the total dollars spent by the customers that the sales support agent provides the service to. If the customers make more purchases, their sales support agent will have a better performance. The total dollars spent by each customer is listed in the *invoice* table. The sales support agents information can be found in the *employee* table.

In [21]:
%%sql
SELECT 
    first_name || ' ' || last_name name, 
    employee_id 
FROM employee 
WHERE title == 'Sales Support Agent'

Done.


[('Jane Peacock', 3), ('Margaret Park', 4), ('Steve Johnson', 5)]

In [22]:
%%sql
SELECT DISTINCT support_rep_id FROM customer

Done.


[(3,), (4,), (5,)]

As shown above, the sales support agents are Jane Peacock, Margaret Park, and Steve Johnson. Their employee_id also match the support_rep_id in the *customer* table.

In [23]:
%%sql
WITH c AS
    (
    SELECT 
        support_rep_id, 
        customer_id
    FROM customer
    )
SELECT 
    e.first_name || ' ' || e.last_name name,
    COUNT(c.customer_id) total_customers,
    ROUND(SUM(i.total), 2) total_dollars_spent,
    ROUND(SUM(i.total)/COUNT(c.customer_id), 2) average_dollars_spent_per_customer
    FROM employee e
    INNER JOIN c ON c.support_rep_id == e.employee_id
    LEFT JOIN invoice i ON i.customer_id == c.customer_id
    GROUP BY 1
    ORDER BY 3 DESC

Done.


[('Jane Peacock', 212, 1731.51, 8.17),
 ('Margaret Park', 214, 1584.0, 7.4),
 ('Steve Johnson', 188, 1393.92, 7.41)]

As shown in the above table, the best-performed sales support agent is Jane Peacock. Margaret Park has slightly less performance, and Steve Johnson has the least performance. Interestingly, the average dollars spent by each customer are the same for Steve Johnson and Margaret Park. However, Margaret Park has a much higher number of customers than Steve Johnson. This indicates that Margaret Park is better at supporting new customers. Jane Peacock has a similar number of customers with Margaret Park, but her customers spend more money on average in-store. It implies that Jane Peacock has a stronger skill at persuading customers to purchase more products. 

### Objective 4: Identify the portions of whole album purchase and individual track purchase
The Chinook store is setup in a way that allows customer to make purchases in one of the two ways:
- purchase a whole album
- purchase a collection of one or more individual tracks

Study the portions of whole album purchase versus individual track purchase can be useful to help the store making management decisions. For example, if the whole album purchase is small, it will be better for the store to purchase the most popular tracks from each album from record companies, instead of purchasing every track from an album. Therefore, my task here is to analyze whether it is a reasonable decision to purchase the most popular tracks. 

To determine whether a purchase is a whole album purchase, I need to identify if the invoice contains all track_id's that an album has. Thus, the columns required are the album_id and the track_id columns from the *track* table, as well as the invoice_id and the track_id columns from the *invoice* table. For each invoice, if the number of tracks from a specific album matches the number of tracks in the album, a whole album must be purchased. I also consider albums that have only one track as individual tracks. 

In [25]:
%%sql
WITH track_by_album AS
    (
    SELECT
        album_id,
        COUNT(track_id) num_of_tracks_album
    FROM track
    GROUP BY 1
    ),
    
    invoice_by_invoice_album AS
    (
    SELECT
        il.invoice_id,
        t.album_id,
        COUNT(t.album_id) num_of_tracks_purchased
    FROM track t
    INNER JOIN invoice_line il ON il.track_id == t.track_id
    GROUP BY 1,2
    ),
    album_purchase AS
    (
    SELECT 
        i.invoice_id,
        i.album_id,
        i.num_of_tracks_purchased,
        t.num_of_tracks_album,
        CASE 
            WHEN i.num_of_tracks_purchased == t.num_of_tracks_album AND t.num_of_tracks_album > 2 THEN 'Yes'
            ELSE 'No'
        END
        AS whole_album_purchase
    FROM track_by_album t
    INNER JOIN invoice_by_invoice_album i ON t.album_id == i.album_id
    GROUP BY invoice_id
    ORDER BY invoice_id
    )
SELECT 
    whole_album_purchase,
    ROUND(CAST(COUNT(*) as Float)/CAST((SELECT COUNT(*) FROM album_purchase) as Float),2) fraction
FROM album_purchase
GROUP BY 1

Done.


[('No', 0.82), ('Yes', 0.18)]

From the table above, 18% of the purchase comes from whole album purchase. Thus, it may not be proper to stop purchasing the whole albums, as it will lose a significant amount of profit. 